mat64/shadow.go

```diff
@@ -223,3 +220,62 @@ func rectanglesOverlap(off, aCols, bCols, stride int) bool {
 	// b strictly wraps and so must overlap with a.
 	return true
 }
+
+// trianglesOverlap returns whether the strided triangles a and b overlap
+// when b is offset by off elements after a but has at least one element before
+// the end of a. a and b have aSize and bSize respectively.
+func trianglesOverlap(off, aSize, bSize, stride int, aUpper, bUpper bool) bool {
+	if !rectanglesOverlap(off, aSize, bSize, stride) {
+		// Fast return if bounding rectangles do not overlap.
+		return false
+	}
+
+	if aUpper {
+		// Test overlap by projecting diagonals from the b upper
+		// left corner, and upper right or lower left corner onto
+		// the top row of a.
+		rows := off / stride
+		proj := off - rows*(stride+1)
+		if proj >= 0 && proj < aSize {
+			// Top left corner of b projects to element in top
+			// row of a.
+			return true
+		}
+		if bUpper {
+			// Project from upper right corner of b.
+			proj += bSize
+			// Return whether corner projects to element in top row of a.
+			return proj >= 0 && proj < aSize
+		}
+		// Since a is an upper triangle and b
+		// is a lower triangle, if the upper
+		// left corner of b is not in a, it is
+		// not possible for there to be an
+		// overlap between the two. This is
+		// for the lower corner of be to be in
+		// a, the data slice of b must have been
+		// before a, but then b would have been
+		// a.
+		return false
+	}
+
+	// Test overlap by projecting diagonals from the b upper
+	// left corner, and upper right or lower left corner onto
+	// the left column of a.
+	cols := off % stride
+	proj := off - cols*(stride+1)
+	if proj >= 0 && proj < aSize*stride {
+		// Top left corner of b projects to element in left
+		// column of a.
+		return true
+	}
+	if bUpper {
```

Comment:
I think that the situation here is the same, a is lower triangle, b is upper triangle, so there can't be overlap if the top left corner is not in a.